tokio/src/runtime/basic_scheduler.rs

```diff
@@ -224,156 +217,102 @@ impl<P: Park> BasicScheduler<P> {
+        let mut park = core.park.take().expect("park missing");
 
-    impl<P: Park> Drop for Guard<'_, P> {
-        fn drop(&mut self) {
-            let Context { tasks, .. } = self.context.take().expect("context missing");
-            self.scheduler.tasks = Some(tasks.into_inner());
-        }
-    }
+        core.stats.submit(&core.spawner.shared.stats);
+        let (mut core, _) = self.enter(core, || {
+            park.park_timeout(Duration::from_millis(0))
+                .expect("failed to park");
+        });
 
-    // Remove `tasks` from `self` and place it in a `Context`.
-    let tasks = scheduler.tasks.take().expect("invalid state");
+        core.park = Some(park);
+        core
+    }
 
-    let guard = Guard {
-        context: Some(Context {
-            shared: scheduler.spawner.shared.clone(),
-            tasks: RefCell::new(tasks),
-        }),
-        scheduler,
-    };
+    fn enter<R>(&self, core: Box<Core>, f: impl FnOnce() -> R) -> (Box<Core>, R) {
+        // Store the scheduler core in the thread-local context
+        //
+        // A drop-guard is employed at a higher level.
+        *self.core.borrow_mut() = Some(core);
 
-    let context = guard.context.as_ref().unwrap();
-    let scheduler = &mut *guard.scheduler;
+        // Execute the closure while tracking the execution budget
+        let ret = f();
 
-    CURRENT.set(context, || f(scheduler, context))
+        // Take the scheduler core back
+        let core = self.core.borrow_mut().take().expect("core missing");
+        (core, ret)
+    }
 }
 
-impl<P: Park> Drop for BasicScheduler<P> {
+impl Drop for BasicScheduler {
     fn drop(&mut self) {
         // Avoid a double panic if we are currently panicking and
         // the lock may be poisoned.
 
-        let mut inner = match self.inner.lock().take() {
-            Some(inner) => inner,
+        let core = match self.take_core() {
+            Some(core) => core,
             None if std::thread::panicking() => return,
-            None => panic!("Oh no! We never placed the Inner state back, this is a bug!"),
+            None => panic!("Oh no! We never placed the Core back, this is a bug!"),
         };
 
-        enter(&mut inner, |scheduler, context| {
+        core.enter(|mut core, context| {
             // Drain the OwnedTasks collection. This call also closes the
             // collection, ensuring that no tasks are ever pushed after this
             // call returns.
-            context.shared.owned.close_and_shutdown_all();
+            context.spawner.shared.owned.close_and_shutdown_all();
 
             // Drain local queue
             // We already shut down every task, so we just need to drop the task.
-            for task in context.tasks.borrow_mut().queue.drain(..) {
-                drop(task);
+            while let Some(task) = core.tasks.pop_front() {
+                let (c, _) = context.enter(core, || drop(task));
+                core = c;
```

Comment:
Here the drop of the task is in the context, but the shutdown isn't. Shouldn't the shutdown also be in the context?

Reply:
Seems like they don't as the scheduler is already shutdown.